tee-worker/ts-tests/utils.ts

```diff
@@ -141,161 +144,47 @@ export async function sendTxUntilInBlock(api: ApiPromise, tx: SubmittableExtrins
-            }
-            console.log(`Chain is at block: #${header.number}`);
-            const signedBlock = await context.substrate.rpc.chain.getBlock(header.hash);
-
-            const allEvents = (await context.substrate.query.system.events.at(header.hash)) as Vec<EventRecord>;
-            signedBlock.block.extrinsics.forEach((ex, index) => {
-                if (
-                    !(
-                        ex.method.section === filterObj.module &&
-                        (ex.method.method === filterObj.method || ex.method.method === 'someError')
-                    )
-                ) {
-                    return;
-                }
-                allEvents
-                    .filter(({ phase, event }) => {
-                        return (
-                            phase.isApplyExtrinsic &&
-                            phase.asApplyExtrinsic.eq(index) &&
-                            event.section == filterObj.module &&
-                            (event.method == filterObj.event || event.method == filterObj.errorEvent)
-                        );
-                    })
-                    .forEach(({ event }) => {
-                        if (event.method == filterObj.event) {
-                            const data = event.data as AESOutput[];
-                            const eventData: HexString[] = [];
-                            for (let i = 0; i < data.length; i++) {
-                                eventData.push(decryptWithAES(aesKey, data[i]));
-                            }
-
-                            resolve(eventData);
-                        } else if (event.method == filterObj.errorEvent) {
-                            const error = event.method;
-
-                            resolve(error);
-                        }
-                        unsubscribe();
-                        return;
-                    });
-            });
-        });
-    });
-}
-export async function listenCreatedIdentityEvents(context: IntegrationTestContext, aesKey: HexString) {
-    return new Promise<{ eventData: HexString[] }>(async (resolve, reject) => {
+// Subscribe to the chain until we get the first specified event with given `section` and `methods`.
+// We can listen to multiple `methods` as long as they are emitted in the same block.
+// The event consumer should do the decryption optionaly as it's event specific
+//
+// TODO: occassionally multiple events for an extrinsic are not included in the same block,
+//       e.g. `create_identity` => `IdentityCreated`, `ChallengeCodeGenerated`
+//       this is because the extrinsics are submitted asynchronously and in rare cases these two
+//       extrinsics are included in the different parentchain blocks
+// Solutions:
+//  1. (pallet change) use one single extrinsic to emit both events, if they should always be triggered on pair
+//  2. (ts-test change) only resolve this promise when both events are received, but not necessarily in the same block
```

Comment:
I think we can simply implement solution 2 like this:
```ts
export async function listenEvent(api: ApiPromise, section: string, methods: string[]) {
    return new Promise<Event[]>(async (resolve, reject) => {
        let startBlock = 0;
        let events: EventRecord[] = []
        const unsubscribe = await api.rpc.chain.subscribeNewHeads(async (header) => {
            const currentBlockNumber = header.number.toNumber();
            if (startBlock == 0) startBlock = currentBlockNumber;
            if (currentBlockNumber > startBlock + listenTimeoutInBlock) {
                reject('timeout');
                return;
            }
            console.log(`\n--------- block #${header.number}, hash ${header.hash} ---------\n`);
            const apiAt = await api.at(header.hash);

            const records = (await apiAt.query.system.events()) as Vec<EventRecord>;
            records.forEach((e, i) => {
                const s = e.event.section;
                const m = e.event.method;
                const d = e.event.data;
                console.log(`Event[${i}]: ${s}.${m} ${d}`);
            });
            const filtered_events = records.filter(
                ({ phase, event }) =>
                    phase.isApplyExtrinsic && section === event.section && methods.includes(event.method)
            );            
            events.push(...filtered_events);
            if (events.length == methods.length) {
                resolve(events.map((e) => e.event));
                unsubscribe();
                return;
            }
        });
    });
}
```

Reply:
Thanks. This however only makes sure the length is equal, but not the content, so theoretically it can be two same events 😂 (if a block contains that).

Anyway I think both option 1 and 2 are needed - they can serve different purposes. I'll leave it to the next pr.

Reply:
```
export async function listenEvent(api: ApiPromise, section: string, methods: string[], txLen: number) {
    return new Promise<Event[]>(async (resolve, reject) => {
        let startBlock = 0;
        let events: EventRecord[] = []
        const unsubscribe = await api.rpc.chain.subscribeNewHeads(async (header) => {
            const currentBlockNumber = header.number.toNumber();
            if (startBlock == 0) startBlock = currentBlockNumber;
            const timeout = await getListenTimeoutInBlocks(api);
            if (currentBlockNumber > startBlock + timeout) {
                reject('timeout');
                return;
            }
            console.log(`\n--------- block #${header.number}, hash ${header.hash} ---------\n`);
            const apiAt = await api.at(header.hash);

            const records: EventRecord[] = (await apiAt.query.system.events()) as any;
            records.forEach((e, i) => {
                const s = e.event.section;
                const m = e.event.method;
                const d = e.event.data;
                console.log(`Event[${i}]: ${s}.${m} ${d}`);
            });
            const filtered_events = records.filter(
                ({ phase, event }) =>
                    phase.isApplyExtrinsic && section === event.section && methods.includes(event.method)
            );
            events.push(...filtered_events)
            if (events.length == txLen) {
                resolve(events.map((e) => e.event));
                unsubscribe();
                return;
            }
        });
    });
}
```
I think it be better to write it like above(I've tested it a few times ).I see this [CI](https://github.com/litentry/litentry-parachain/actions/runs/4441086313/jobs/7797592061?pr=1477) failed,maybe trigger this, I remember this pr and `TODO`. txLen is [txs.length](https://github.com/litentry/litentry-parachain/blob/dev/tee-worker/ts-tests/indirect_calls.ts#L49). cc @BillyWooo